talk/morelanguage/exceptions.tex

```diff
@@ -4,185 +4,166 @@
   \begin{block}{}
-    After a lot of thinking and experiencing, the conclusions of the community on exception handling are :
     \begin{itemize}
-    \item Never write an exception specification
-    \item Except possibly an empty one
-    \end{itemize}
-  \end{block}
-  \pause
-  \begin{alertblock}{Some of the reasons}
-    \begin{itemize}
-    \item throw specification is runtime only
+      \item exceptions have little cost if no exception is thrown
       \begin{itemize}
-      \item does not allow compiler optimizations
-      \item on the contrary forces extra checks
-      \item generally terminates your program if violated
+        \item a throw usually costs about as much as a function call
+        \item so minimal if you are using exceptions for \textit{exceptional} events
+        \item thus their use is generally recommended
       \end{itemize}
-    \item throw specification clashes with templates
+      \item however, in a tight, performance critical loop, where the error is handled close to the place it happens a C-style error code might be better
       \begin{itemize}
-      \item one cannot ``template'' the throw clause
+        \item But profile if you really want to find out!
       \end{itemize}
-    \end{itemize}
-  \end{alertblock}
-\end{frame}
-
-\begin{frame}[fragile]
-  \frametitlecpp[11]{What remains}
-  \begin{block}{throw is dead}
-    \begin{itemize}
-    \item dynamic exception specification are deprecated
-    \item even throw() (no exceptions)
-    \end{itemize}
+   \end{itemize}
   \end{block}
-  \pause
-  \begin{exampleblock}{long live noexcept}
-    \begin{itemize}
-    \item noexcept a somehow equivalent to throw()
-    \item but is checked at compile time
-    \item so allows compiler optimizations
-    \end{itemize}
-  \end{exampleblock}
+  \begin{multicols}{2}
+    \begin{cppcode*}{fontsize=\tiny,gobble=2}
+      // try ... catch in hot loop
+      //   might impact performance
+      for (const &item: list) {
+        try {
+          check_item(item); // Can throw bad_item
+          ... // process item
+        }
+        catch (bad_item &e) {
+          ... // ignore this item, but keep going
+        }
+      }
+    \end{cppcode*}
+    \columnbreak
+    \begin{cppcode*}{fontsize=\tiny,gobble=2}
+      // For this code logic C-style error codes
+      //   may be better
+      for (const &item: list) {
+        auto err_code = check_item(item);
+        if (!err_code) {
+          ... // process item
+        } else {
+          ... // warn and carry on
+        }
+      }
+    \end{cppcode*}
+  \end{multicols}
 \end{frame}
 
+
 \begin{frame}[fragile]
-  \frametitlecpp[11]{Full power of noexcept}
-  \begin{block}{3 uses of noexcept}
+  \frametitlecpp[11]{noexcept}
+  \begin{block}{noexcept}
     \begin{itemize}
-    \item standalone
-      \begin{cppcode*}{gobble=2, linenos=false}
+      \item specifying the \textit{noexcept} operator states that a function is guaranteed not to throw an exception
```

Comment:
```suggestion
      \item specifying the \textit{noexcept} specifier states that a function is guaranteed not to throw an exception
```
The `noexcept` operator (written the same way) is used to query that property from a function, like `bool b = noexcept(f(2, 54));`

Reply:
Yep, honestly I'd say this is rather an advanced topic (I was indeed confused). I have tried to simplify this to essentials that are useful now.